src/Psy/ExecutionLoop/Loop.php

```diff
@@ -27,17 +30,55 @@ class Loop
      */
     public function __construct(Configuration $config)
     {
-        // don't need this
+        $this->cleaner = $config->getCodeCleaner();
     }
 
     /**
-     * Run the execution loop.
-     *
-     * @param Shell $shell
+     * @return callable
      */
-    public function run(Shell $shell)
+    protected function getExecutionClosure()
     {
-        $loop = function ($__psysh__) {
+        return function (Shell &$__psysh__, $code = null) {
+
+            try {
+                extract($__psysh__->getScopeVariables());
+
+                // evaluate the current code buffer
+                ob_start();
+
+                set_error_handler(array($__psysh__, 'handleError'));
+                $_ = eval(!is_null($code) ? $code : $__psysh__->flushCode());
+                restore_error_handler();
+
+                $__psysh_out__ = ob_get_contents();
+                ob_end_clean();
+
+                $__psysh__->writeStdout($__psysh_out__);
+                $__psysh__->writeReturnValue($_);
+
+                // a bit of housekeeping
+                unset($__psysh_out__);
+                unset($code);
+
+                $__psysh__->setScopeVariables(get_defined_vars());
+            } catch (BreakException $_e) {
```

Comment:
For example, handling BreakException only makes sense if you could be executing commands, not if you're evaluating code.